bauble/src/value.rs

```diff
@@ -1082,17 +1109,111 @@ impl<'a> Symbols<'a> {
     }
 }
 
+/// We can delay registering `Ref` assets if what they're referencing hasn't been loaded yet.
+pub struct DelayedRegister {
+    pub asset: Spanned<TypePath>,
+    pub reference: Spanned<PathKind>,
+}
+
+pub fn resolve_delayed(
+    mut delayed: Vec<DelayedRegister>,
+    ctx: &mut crate::context::BaubleContext,
+) -> std::result::Result<(), Vec<Spanned<ConversionError>>> {
+    loop {
+        let old_len = delayed.len();
+
+        // Try to register delayed registers, and remove them as they succeed.
+        delayed.retain(|d| {
+            if let Some(r) = match &d.reference.value {
+                PathKind::Direct(path) => ctx.get_ref(path.borrow()),
+                PathKind::Indirect(path, ident) => {
+                    ctx.ref_with_ident(path.borrow(), ident.borrow())
+                }
+            } && let Some((ty, _)) = &r.asset
+            {
+                ctx.register_asset(d.asset.value.borrow(), *ty);
+                false
+            } else {
+                true
+            }
+        });
+
+        if delayed.is_empty() {
+            return Ok(());
+        }
+
+        // If the length didn't change we have errors.
+        if delayed.len() == old_len {
+            let mut graph = petgraph::graphmap::DiGraphMap::new();
+            let mut map = HashMap::new();
+            for a in delayed.iter() {
+                let node_a = graph.add_node(a.asset.as_ref().map(|p| p.borrow()));
+                map.insert(node_a, a.reference.as_ref());
+
+                for b in delayed.iter() {
+                    if a.reference.could_be(b.asset.borrow()) {
+                        graph.add_edge(node_a, b.asset.as_ref().map(|p| p.borrow()), ());
+                    }
+                }
+            }
+
+            let mut errors = Vec::new();
+            for scc in petgraph::algo::tarjan_scc(&graph) {
+                if scc.len() == 1 {
```

Comment:
I mean this is done in a lot of places throughout the code. I understand it means the path is an identifier, but it doesn't exactly make sense to me why it has to be an identifier here yet again? Like, what is wrong with it being a fully qualified path?

Reply:
That's not a path? `tarjan_scc` returns all nodes as cycles in a graph. So nodes that aren't in a cycle it returns an array of length 1 for.